scripts/pretraining/bert/create_pretraining_data.py

```diff
@@ -0,0 +1,532 @@
+        logging.debug('*** Writing to output file %s ***', output_file)
+        write_to_files_np(features, tokenizer, max_seq_length,
+                          max_predictions_per_seq, [output_file])
+        features = None
+    else:
+        features = (input_ids, masked_lm_ids, masked_lm_positions, masked_lm_weights,
+                    next_sentence_labels, segment_ids, valid_lengths)
+    time_end = time.time()
+    logging.debug('Process %d files took %.1f s', len(input_files), time_end - time_start)
+    return features
+
+
+def create_instances_from_document(x):
+    """Creates `TrainingInstance`s for a single document."""
+    (all_documents, document_index, max_seq_length, short_seq_prob,
+     masked_lm_prob, max_predictions_per_seq, whole_word_mask, vocab, tokenizer) = x
+    document = all_documents[document_index]
+    _MASK_TOKEN = vocab.mask_id
+    _CLS_TOKEN = vocab.cls_id
+    _SEP_TOKEN = vocab.sep_id
+
+    # Account for [CLS], [SEP], [SEP]
+    max_num_tokens = max_seq_length - 3
+
+    # According to the original tensorflow implementation:
+    # We *usually* want to fill up the entire sequence since we are padding
+    # to `max_seq_length` anyways, so short sequences are generally wasted
+    # computation. However, we *sometimes*
+    # (i.e., short_seq_prob == 0.1, 10% of the time) want to use shorter
+    # sequences to minimize the mismatch between pre-training and fine-tuning.
+    # The `target_seq_length` is just a rough target however, whereas
+    # `max_seq_length` is a hard limit.
+    target_seq_length = max_num_tokens
+    if random.random() < short_seq_prob:
+        target_seq_length = random.randint(2, max_num_tokens)
+
+    # We DON'T just concatenate all of the tokens from a document into a long
+    # sequence and choose an arbitrary split point because this would make the
+    # next sentence prediction task too easy. Instead, we split the input into
+    # segments "A" and "B" based on the actual "sentences" provided by the user
+    # input.
+    instances = []
+    current_chunk = []
+    current_length = 0
+    i = 0
+    while i < len(document):  # pylint: disable=R1702
+        segment = document[i]
+        current_chunk.append(segment)
+        current_length += len(segment)
+        if i == len(document) - 1 or current_length >= target_seq_length:
+            if current_chunk:
+                # `a_end` is how many segments from `current_chunk` go into the `A`
+                # (first) sentence.
+                a_end = 1
+                if len(current_chunk) >= 2:
+                    a_end = random.randint(1, len(current_chunk) - 1)
+
+                tokens_a = []
+                for j in range(a_end):
+                    tokens_a.extend(current_chunk[j])
+
+                tokens_b = []
+                # Random next
+                is_random_next = False
+                if len(current_chunk) == 1 or random.random() < 0.5:
```

Comment:
You may try to implement the sentence order prediction objective in ALBERT. See https://github.com/google-research/albert/blob/cbce824721148a8eb2621f1b1c1af252ac5121cc/create_pretraining_data.py#L338-L345

Reply:
done